src/driver-sinowealth-nubwo.c

```diff
@@ -0,0 +1,315 @@
+#include <stdint.h>
+#include <string.h>
+
+#define SINOWEALTHNUBWO_PERF_CMD_REPORTID 0x02
+#define SINOWEALTHNUBWO_AESTHETIC_CMD_REPORTID 0x03
+#define SINOWEALTHNUBWO_GET_FIRMWARE_CMD_REPORTID 0x04
+
+#define SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE 256
+#define SINOWEALTHNUBWO_GET_FIRMWARE_MSGOFFSET 48
+
+#define SINOWEALTHNUBWO_PERF_CMD_MSGSIZE 16
+
+// Actually more but I only implemented one
+#define SINOWEALTHNUBWO_NUM_PROFILES 1
+#define SINOWEALTHNUBWO_NUM_RESOLUTIONS 1
+// Actually 8 but I am not going to implement macros.
+#define SINOWEALTHNUBWO_NUM_BUTTONS 0
+#define SINOWEALTHNUBWO_NUM_LEDS 1
+
+//Magic set_feature that must be called before requesting firmware string
+static uint8_t PREFIRMWARE_QUERY_MSG[] = {0x02, 0x01, 0x49, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+static unsigned int REPORT_RATES[] = { 125, 250, 333, 500, 1000 };
+static uint8_t REPORT_RATES_ENCODED[] = {0x08, 0x04, 0x03, 0x02, 0x01};
+static uint8_t REPORT_RATES_CMD[] = {0x02, 0x06, 0xbb, 0xaa, 0x28, 0x00, 0x01, 0x00};
+
+static unsigned int DPILIST[] = { 1000, 2000, 3000, 5000, 15000};
+static uint8_t DPI_ENCODED[] = {0x04, 0x03, 0x02, 0x01, 0x00};
+static uint8_t DPI_CMD[] = { 0x02, 0x06, 0xbb, 0xaa, 0x32, 0x00, 0x01, 0x00 };
+
+enum
+sinowealthnubwo_color_mode {
+	SINOWEALTHNUBWO_COLOR_OFF = 0x00,
+	SINOWEALTHNUBWO_COLOR_ON = 0x01,
+	SINOWEALTHNUBWO_COLOR_BREATHING = 0x02,
+	SINOWEALTHNUBWO_COLOR_COLOR_SHIFT = 0x03,
+	SINOWEALTHNUBWO_COLOR_SPECTRUM = 0x04,
+	SINOWEALTHNUBWO_COLOR_MARQUEE = 0x05,
+};
+
+struct
+sinowealthnubwo_aesthetic_report {
+	uint8_t report_id; //0x03
+	uint8_t cmd[7];
+	uint8_t r;
+	uint8_t g;
+	uint8_t b;
+	uint8_t color_mode;
+	uint8_t padzero; // 0x00
+	uint8_t brightness; // 0x01 to 0x03
+	uint8_t tempo; // 0x05 0x03 0x01
+	uint8_t padzero2[16*3+1];
+} __packed__;
+
+static uint8_t AESTHETIC_CMD[] = {0x06, 0xbb, 0xaa, 0x2a, 0x00, 0x0a, 0x00};
+
+static int
+sinowealthnubwo_test_hidraw(struct ratbag_device *device) 
+{
+	return 
+		ratbag_hidraw_has_report(device, SINOWEALTHNUBWO_AESTHETIC_CMD_REPORTID)
+		&& ratbag_hidraw_has_report(device, SINOWEALTHNUBWO_PERF_CMD_REPORTID)
+		&& ratbag_hidraw_has_report(device, SINOWEALTHNUBWO_GET_FIRMWARE_CMD_REPORTID);
+}
+
+static char*
+sinowealth_get_firmware_string(struct ratbag_device *device)
+{
+	uint8_t buff[SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE];
+	int size;
+
+	size = ratbag_hidraw_set_feature_report(device, SINOWEALTHNUBWO_PERF_CMD_REPORTID, PREFIRMWARE_QUERY_MSG, ARRAY_LENGTH(PREFIRMWARE_QUERY_MSG));
+	if (size < 0)
+		log_error(device->ratbag, "Error while sending pre-firmware request message: %d\n", size);
+
+	size = ratbag_hidraw_get_feature_report(device, SINOWEALTHNUBWO_GET_FIRMWARE_CMD_REPORTID, buff, SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE);
+	if (size != SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE)
+		log_error(device->ratbag ,"Firmware report reply size mismatch expected %d got %d\n", SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE, size);
```

Comment:
In both of these `log_error` cases you might want to return a `strdup`'d constant, since log_error just emits the log message, it doesn't exit/panic the process or anything like that.